File: 7-kyu/ga-de-ry-po-lu-ki-cypher/c++/solution.cpp
#include <cctype>
#include <map>
#include <string>

std::map<char, char> lookup_table = {
    { 'g', 'a' },
    { 'a', 'g' },
    { 'd', 'e' },
    { 'e', 'd' },
    { 'r', 'y' },
    { 'y', 'r' },
    { 'p', 'o' },
    { 'o', 'p' },
    { 'l', 'u' },
    { 'u', 'l' },
    { 'k', 'i' },
    { 'i', 'k' }
};

std::string encode (const std::string& str)
{
    auto str_cpy = str;
    for (char &c : str_cpy) {
        bool upperflag = std::isupper(c);
        char lower_c = upperflag ? std::tolower(c) : c;
        if (lookup_table[lower_c]) {
            lower_c = lookup_table[lower_c];
            c = upperflag ? std::toupper(lower_c) : lower_c;
        }
    }
    return str_cpy;
}

std::string decode (const std::string& str)
{
    return encode(str);
}

File: 7-kyu/ga-de-ry-po-lu-ki-cypher/python/solution.py
LOOKUP_TABLE = {'g': 'a', 'a': 'g', 'd': 'e', 'e': 'd', 'r': 'y',
                'y': 'r', 'p': 'o', 'o': 'p', 'l': 'u', 'u': 'l',
                'k': 'i', 'i': 'k'
               }

def encode(message):
    ret = ''
    for ch in message:
        if ch.lower() in LOOKUP_TABLE.keys():
            if ch.islower():
                ret += LOOKUP_TABLE[ch]
            else:
                ret += LOOKUP_TABLE[ch.lower()].upper()
        else:
            ret += ch
    return ret

def decode(message):
    return encode(message)
